cpp/ql/lib/semmle/code/cpp/commons/StringConcatenation.qll

```diff
@@ -0,0 +1,98 @@
+/**
+ * A library for detecting general string concatenations.
+ */
+
+import cpp
+import semmle.code.cpp.models.implementations.Strcat
+import semmle.code.cpp.models.interfaces.FormattingFunction
+import semmle.code.cpp.dataflow.new.DataFlow
+
+class StringConcatenation extends Call {
+  StringConcatenation() {
+    // sprintf-like functions, i.e., concat through formating
+    exists(FormattingFunctionCall fc | this = fc)
+    or
+    this.getTarget() instanceof StrcatFunction
+    or
+    this.getTarget() instanceof StrlcatFunction
+    or
+    // operator+ concat
+    exists(Call call, Operator op |
+      call.getTarget() = op and
+      op.hasQualifiedName(["std", "bsl"], "operator+") and
+      op.getType().(UserType).hasQualifiedName(["std", "bsl"], "basic_string") and
+      this = call
+    )
+    or
+    // string stream concat (operator<<)
+    this.getTarget().hasQualifiedName(["std", "bsl"], "operator<<")
+  }
+
+  /**
+   * Gets the operands of this concatenation (one of the string operands being
+   * concatenated).
+   * Will not return out param for sprintf-like functions, but will consider the format string
+   * to be part of the operands.
+   */
+  Expr getAnOperand() {
+    // The result is an argument of 'this' (a call)
+    result = this.getAnArgument() and
+    // addresses odd behavior with overloaded operators
+    // i.e., "call to operator+" appearing as an operand
```

Comment:
Is this where you have more than one concatenation, e.g. `"a" + "b" + "c"`?  I think in that case we would expect one of the concatenations to have the other (a "call to operator+") as an argument.

Reply:
Spoke to @MathiasVP , I'll spell it out in the comments as to why I chose this approach and why it is necessary.